### PR TITLE
lite: detach model_output tensors before stashing per-microbatch on cp=0

### PR DESCRIPTION
**Pull request by ISEEKYAN**
Port of **verl-project/Megatron-LM#32** by @shyoshyo (leikaixiang.shyoshyo@bytedance.com). Original commit cherry-picked as-is; authorship and commit message preserved, no squash, no rewrite.

## What it fixes

On the last PP stage, only `is_mp_src_rank_with_outputs()` (tp==0 && cp==0 && pp==last) allocates the runtime `output_lst`, and the loss hook appends one `{"model_output": ..., "loss": ..., "metrics": ...}` per microbatch. `model_output["log_probs"]` comes from the protocol's `unpack_forward_output`, which on CP>1 chains through the differentiable `torch.distributed.nn.functional.all_gather` -> `torch.cat` -> slice -> `torch.nested.as_nested_tensor`. Those tensors keep `requires_grad=True` and are *not* on the path `loss.mean().backward()` traverses, so their saved-for-backward buffers stay pinned for the whole step.

Upstream reproduced this on DS-V4-Flash (PP=4 EP=2 CP=2, MAX_LENGTH=16384, 128 microbatches/step): rank=6 (cp=0) grows ~0.5 GB per microbatch to ~60 GB while rank=7 (cp=1) stays flat at 4.45 GB; also reproduced at PP=4 EP=8 CP=8, 32768. Fits in HBM at 16k, OOMs deterministically at 128k.

The fix detaches every tensor in `model_output` before stashing the per-microbatch copy. The live `model_output` handed to `loss_function` is untouched, so gradients are unchanged.

## Review notes (checked against this repo's `main`)

- **Detach is semantics-preserving here.** `_build_verl_model_output` only ever returns `log_probs` and (optionally) `entropy`, both plain tensors, so the dict comprehension covers the whole payload -- no list/tuple/nested container escapes it.
- **No consumer needs the graph.** `output_lst` is handed to `postprocess_batch_func` only *after* `runtime.forward_backward(...)` returns, i.e. after backward has already run; differentiating through those tensors was never possible. `forward_only` paths need no grad at all.
- **Consistent with the dynamic-CP path**, whose `_record_output` in `dynamic_cp.py` already does the equivalent `.detach().cpu()`. This brings the static path in line.
- Diff is byte-identical to the upstream PR head (`3282f104`) and applies cleanly on `main`.
- Changed lines are clean under the repo's `experimental/lite/.pre-commit-config.yaml` toolchain (isort + black 24.4.2 with `--skip-magic-trailing-comma --skip-string-normalization`); the only formatting/lint noise in this file is pre-existing on `main` and untouched here.

## Not covered

No regression test is included -- reproducing the leak needs a multi-GPU CP>1 run. A cheap follow-up would be a unit assertion that every tensor collected into `output_lst` has `requires_grad=False`. Upstream's evidence is an end-to-end training run (memory curves + loss/metric parity spot check), not an in-tree test.

Upstream PR: https://github.com/verl-project/Megatron-LM/pull/32

### CODE REVIEW

**Comment by ISEEKYAN** [comment]
**Test status: this port has NOT been validated by any executed test.** Recording it here so the state is not overstated.

The repo's only sanctioned entry point, `experimental/lite/tests/run_tests.sh`, is hermetic: `_test_harness/runner.py` overwrites `PYTHONPATH` with `REPO_ROOT:LITE_ROOT`. The three tests that cover the changed file (`unit/verl/test_mlite_engine_metrics.py`, `unit/verl/test_mlite_engine_loss_mask_packing.py`, `unit/examples/test_verl_mlite_engine_loss_mask_packing.py`) import `MegatronLiteEngine`, which needs `verl` + `transformer_engine`. Our cluster images are RL-stack images whose `verl`/`transformers` come from overlays injected via `PYTHONPATH` -- exactly what the hermetic runner strips. Grafting the overlay packages into `site-packages` gets one dependency further each time and then hits the next version pin (`transformers` v5 vs verl's v4 API; then `huggingface-hub` 1.19 vs `<1.0`).

One useful control did come out of it: running those three files with the production overlay chain under plain `pytest` (not the sanctioned entry point) gives **byte-identical collection failures on this branch and on unpatched `main`** -- so nothing here is caused by this change. Separately, `unit/model/deepseek_v4/*` cannot collect on `main` either (`cannot import name 'csa_cp_layout_kernels'`), also unrelated.

Getting real coverage needs the dev container from `docker/Dockerfile.ci.dev` per `experimental/lite/tests/README.md`, which we do not currently have built. Tracked separately.

Review basis for this PR therefore remains code reading, plus upstream's end-to-end evidence in verl-project/Megatron-LM#32.

**Comment by ISEEKYAN** [comment]
**Update: this port is now validated. The earlier "not validated by any executed test" comment above is superseded.**

Ran the repo's sanctioned entry point (`experimental/lite/tests/run_tests.sh`, no arguments -> `profile=standard`) on one 8xH100 node, two arms in the same container: this branch (`8be59b0`) and unpatched `main` (`1407993`). Slurm job `16029014`.

Twelve suites PASS, **byte-identical between the two arms**:

```
suite=cpu                            PASS  701 passed / 0 failed / 0 skipped
suite=1gpu-csa-thd-cp                PASS  3
suite=1gpu-fsdp2-offload-gpu         PASS  2
suite=1gpu-fsdp2-unit                PASS  1
suite=1gpu-glm5-lite-static          PASS  3 + 4
suite=1gpu-hf-weights-gpu            PASS  3
suite=1gpu-hf-weights-streaming      PASS  1
suite=1gpu-qwen-lite-forward-smoke   PASS  2
suite=1gpu-router-aux-loss-gate      PASS  2
suite=2gpu-distopt-checkpoint-smoke  PASS  1
suite=2gpu-fsdp2-offload-checkpoint  PASS  3
suite=2gpu-glm5-lite-cp-smoke        FAIL  2 passed / 3 failed   <-- identical on unpatched main
```

The three files that cover the changed code (`unit/verl/test_mlite_engine_metrics.py`, `unit/verl/test_mlite_engine_loss_mask_packing.py`, `unit/examples/test_verl_mlite_engine_loss_mask_packing.py`) give `overall=PASS exit_code=0` on both arms.

**Honest scope of that evidence:** it establishes *no regression*, not *the leak is fixed*. Those suites exercise the engine's metric and loss-mask-packing paths; none of them measures per-microbatch activation memory. The proof that the leak is gone remains upstream's end-to-end memory curves in verl-project/Megatron-LM#32.

Two things intentionally left outside this PR:

1. `2gpu-glm5-lite-cp-smoke` is red on `main` already (`test_glm5_dsa_cp2_*`, `test_glm5_tiny_model_cp2_*`). Because the harness stops at the first non-PASS suite, the 8-GPU suite (`smoke/workflows/checkpoint/test_save_load_export_smoke.py`, 21 cases) is still never reached.
2. `tests/unit/primitive/ckpt/test_hf_weights_gpu.py` is missing its `gpus` marker, so the harness classifies it as a CPU test, its five CUDA tests skip, and skip-is-not-pass fails the whole `cpu` suite -- which is why every suite after it had never run. Verified fix (module-level `gpus(1)` plus per-test `gpus(2)` on the two `_require_two_ranks` tests) is going out as its own PR.
